src/Workspaces/Core/Portable/Workspace/Workspace_Editor.cs

```diff
@@ -455,16 +444,43 @@ private ISet<DocumentId> GetProjectOpenDocuments_NoLock(ProjectId project)
                 var newDoc = newSolution.GetDocument(documentId);
                 this.OnDocumentTextChanged(newDoc);
 
-                // Fire and forget that the workspace is changing
-                var ignore1 = RaiseWorkspaceChangedEventAsync(WorkspaceChangeKind.DocumentChanged, oldSolution, newSolution, documentId: documentId);
-                var ignore2 = this.RaiseDocumentOpenedEventAsync(newDoc); // don't await this
+                // Fire and forget that the workspace is changing.
+                RaiseWorkspaceChangedEventAsync(WorkspaceChangeKind.DocumentChanged, oldSolution, newSolution, documentId: documentId);
+                this.RaiseDocumentOpenedEventAsync(newDoc);
+            }
+
+            this.RegisterText(textContainer);
+        }
+
+        private class ReuseVersionLoader : TextLoader
+        {
+            // Capture DocumentState instead of Document so that we don't hold onto the old solution.
```

Comment:
Should this just hold onto the ValueSource<TextAndVersion> so it's very clear it's not using anything else from DocumentState? I'm still a bit worried this would be more expensive than we want.